File: DeenTime/backend/DeenTime.Api/Controllers/IqamaController.cs
using DeenTime.Core.Entities;
using DeenTime.Core.Enums;
using DeenTime.Infrastructure;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using DeenTime.Api.Requests.Iqama;
using DeenTime.Api.Authorization;

namespace DeenTime.Api.Controllers
{
	[ApiController]
	[Authorize]
	[Route("api/v1/[controller]")]
	public sealed class IqamaController : ControllerBase
	{
		private readonly AppDbContext _db;
		public IqamaController(AppDbContext db) { _db = db; }

		[HttpGet]
		public async Task<IActionResult> List([FromQuery] Guid orgId, [FromQuery] int year)
		{
			if (!User.CanAccessOrganization(orgId)) return Forbid();
			var items = await _db.IqamaEntries.AsNoTracking()
				.Where(i => i.OrganizationId == orgId && i.Date.Year == year)
				.OrderBy(i => i.Date).ThenBy(i => i.Salah).ToListAsync();
			return Ok(items);
		}

		[HttpGet("current")]
		public async Task<IActionResult> Current(
			[FromQuery] Guid orgId,
			[FromQuery] DateOnly? date,
			CancellationToken cancellationToken)
		{
			if (!User.CanAccessOrganization(orgId)) return Forbid();
			var effectiveDate = date ?? DateOnly.FromDateTime(DateTime.UtcNow);
			var history = await _db.IqamaEntries.AsNoTracking()
				.Where(entry => entry.OrganizationId == orgId && entry.Date <= effectiveDate)
				.OrderBy(entry => entry.Date)
				.ThenBy(entry => entry.UpdatedAtUtc)
				.ToListAsync(cancellationToken);

			var current = history
				.GroupBy(entry => entry.Salah)
				.Select(group => group.Last())
				.OrderBy(entry => SalahOrder(entry.Salah))
				.ToArray();
			return Ok(current);
		}

		[HttpPut("schedule")]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> UpdateSchedule(
			[FromBody] IqamaScheduleUpsertRequest req,
			CancellationToken cancellationToken)
		{
			if (!User.CanAccessOrganization(req.OrganizationId)) return Forbid();
			if (req.Entries is null || req.Entries.Length == 0)
				return BadRequest("Add at least one Iqama time.");
			if (req.Entries.Length > 9)
				return BadRequest("A schedule can contain at most nine prayer entries.");
			if (req.Entries.GroupBy(entry => entry.Salah).Any(group => group.Count() > 1))
				return BadRequest("Each prayer can appear only once for an effective date.");
			if (req.Entries.Any(entry => entry.OffsetMinutes is < 0 or > 180))
				return BadRequest("Minutes after prayer start must be between 0 and 180.");

			var salahs = req.Entries.Select(entry => entry.Salah).ToArray();
			var existing = await _db.IqamaEntries
				.Where(entry => entry.OrganizationId == req.OrganizationId &&
					entry.Date == req.EffectiveDate && salahs.Contains(entry.Salah))
				.ToDictionaryAsync(entry => entry.Salah, cancellationToken);

			foreach (var item in req.Entries)
			{
				if (!existing.TryGetValue(item.Salah, out var entity))
				{
					entity = new IqamaEntry
					{
						Id = Guid.NewGuid(),
						OrganizationId = req.OrganizationId,
						Date = req.EffectiveDate,
						Salah = item.Salah
					};
					_db.IqamaEntries.Add(entity);
				}

				entity.Time = item.Time;
				entity.OffsetMinutes = item.OffsetMinutes;
				entity.Note = string.IsNullOrWhiteSpace(item.Note) ? null : item.Note.Trim();
				entity.UpdatedAtUtc = DateTime.UtcNow;
			}

			await _db.SaveChangesAsync(cancellationToken);
			var saved = req.Entries
				.Select(item => existing.GetValueOrDefault(item.Salah) ??
					_db.IqamaEntries.Local.First(entry => entry.OrganizationId == req.OrganizationId &&
						entry.Date == req.EffectiveDate && entry.Salah == item.Salah))
				.OrderBy(entry => SalahOrder(entry.Salah))
				.ToArray();
			return Ok(saved);
		}

		[HttpPost]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> Create([FromBody] IqamaUpsertRequest req)
		{
			if (!User.CanAccessOrganization(req.OrganizationId)) return Forbid();
			if (req.OffsetMinutes is < 0 or > 180) return BadRequest("Minutes after prayer start must be between 0 and 180.");
			if (await _db.IqamaEntries.AnyAsync(entry => entry.OrganizationId == req.OrganizationId && entry.Date == req.Date && entry.Salah == req.Salah))
				return Conflict("An Iqama entry already exists for that prayer and effective date.");
			var entity = new IqamaEntry
			{
				Id = Guid.NewGuid(), OrganizationId = req.OrganizationId, Date = req.Date,
				Salah = req.Salah, Time = req.Time, Note = req.Note, OffsetMinutes = req.OffsetMinutes
			};
			_db.IqamaEntries.Add(entity);
			await _db.SaveChangesAsync();
			return Created($"/api/v1/iqama/{entity.Id}", entity);
		}

		[HttpPut("{id:guid}")]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> Update(Guid id, [FromBody] IqamaUpsertRequest req)
		{
			var existing = await _db.IqamaEntries.FirstOrDefaultAsync(i => i.Id == id);
			if (existing is null) return NotFound();
			if (!User.CanAccessOrganization(existing.OrganizationId) || req.OrganizationId != existing.OrganizationId) return Forbid();
			if (req.OffsetMinutes is < 0 or > 180) return BadRequest("Minutes after prayer start must be between 0 and 180.");
			existing.Date = req.Date; existing.Salah = req.Salah;
			existing.Time = req.Time; existing.Note = req.Note; existing.UpdatedAtUtc = DateTime.UtcNow;
			existing.OffsetMinutes = req.OffsetMinutes;
			await _db.SaveChangesAsync();
			return NoContent();
		}

		[HttpDelete("{id:guid}")]
		[Authorize(Roles = "Admin,Editor")]
		public async Task<IActionResult> Delete(Guid id)
		{
			var existing = await _db.IqamaEntries.FirstOrDefaultAsync(i => i.Id == id);
			if (existing is null) return NotFound();
			if (!User.CanAccessOrganization(existing.OrganizationId)) return Forbid();
			_db.IqamaEntries.Remove(existing);
			await _db.SaveChangesAsync();
			return NoContent();
		}

		private static int SalahOrder(SalahType salah) => salah switch
		{
			SalahType.Fajr => 0,
			SalahType.Dhuhr => 1,
			SalahType.Asr => 2,
			SalahType.Maghrib => 3,
			SalahType.Isha => 4,
			SalahType.Jumuah => 5,
			SalahType.Jumuah2nd => 6,
			SalahType.Jumuah3rd => 7,
			SalahType.Jumuah4th => 8,
			_ => 99
		};
	}
}

File: DeenTime/backend-node/src/modules/iqama/iqama.controller.ts
import { Body, Controller, Delete, Get, HttpCode, Param, Post, Put, Query, Res } from '@nestjs/common';
import type { Response } from 'express';
import { randomUUID } from 'node:crypto';
import { AuthorizeRoles, Authorize, CurrentUser } from '../../common/auth/authorize.decorator.js';
import { canAccessOrganization } from '../../common/auth/claims.js';
import type { SessionUser } from '../../common/auth/jwt.service.js';
import { badRequestText, conflict, forbidden, notFound } from '../../common/errors.js';
import { isGuid, parseDateOnly, salahToDb, timeOnlyToDb, utcToday } from '../../common/json.js';
import { validated } from '../../common/zod-validation.pipe.js';
import { PrismaService } from '../../prisma/prisma.service.js';
import { latestEntriesPerSalah, salahOrder, toIqamaEntryJson, type IqamaEntryJson } from './iqama.mapper.js';
import { iqamaScheduleSchema, iqamaUpsertSchema, type IqamaScheduleRequest, type IqamaUpsertRequest } from './iqama.schemas.js';

const OFFSET_RANGE_MESSAGE = 'Minutes after prayer start must be between 0 and 180.';

/** Port of IqamaController.cs. */
@Controller('api/v1/iqama')
@Authorize()
export class IqamaController {
  constructor(private readonly prisma: PrismaService) {}

  @Get()
  async list(@CurrentUser() user: SessionUser | null, @Query('orgId') orgId?: string, @Query('year') year?: string): Promise<IqamaEntryJson[]> {
    const organizationId = requireGuid(orgId);
    if (!canAccessOrganization(user, organizationId)) throw forbidden();
    const yearNumber = Number.parseInt(year ?? '', 10) || 0;
    const rows = await this.prisma.iqamaEntry.findMany({
      where: { organizationId, date: { gte: new Date(Date.UTC(yearNumber, 0, 1)), lt: new Date(Date.UTC(yearNumber + 1, 0, 1)) } },
      orderBy: [{ date: 'asc' }, { salah: 'asc' }],
    });
    return rows.map(toIqamaEntryJson);
  }

  @Get('current')
  async current(@CurrentUser() user: SessionUser | null, @Query('orgId') orgId?: string, @Query('date') date?: string): Promise<IqamaEntryJson[]> {
    const organizationId = requireGuid(orgId);
    if (!canAccessOrganization(user, organizationId)) throw forbidden();
    const effectiveDate = (date ? parseDateOnly(date) : null) ?? utcToday();
    const history = await this.prisma.iqamaEntry.findMany({
      where: { organizationId, date: { lte: effectiveDate } },
      orderBy: [{ date: 'asc' }, { updatedAtUtc: 'asc' }],
    });
    return latestEntriesPerSalah(history)
      .map(toIqamaEntryJson)
      .sort((a, b) => salahOrder(a.salah) - salahOrder(b.salah));
  }

  @Put('schedule')
  @AuthorizeRoles('Admin', 'Editor')
  async updateSchedule(@CurrentUser() user: SessionUser | null, @Body(validated(iqamaScheduleSchema)) req: IqamaScheduleRequest): Promise<IqamaEntryJson[]> {
    if (!canAccessOrganization(user, req.organizationId)) throw forbidden();
    if (req.entries.length === 0) throw badRequestText('Add at least one Iqama time.');
    if (req.entries.length > 9) throw badRequestText('A schedule can contain at most nine prayer entries.');
    if (new Set(req.entries.map((entry) => entry.salah)).size !== req.entries.length) throw badRequestText('Each prayer can appear only once for an effective date.');
    if (req.entries.some((entry) => entry.offsetMinutes !== null && (entry.offsetMinutes < 0 || entry.offsetMinutes > 180))) throw badRequestText(OFFSET_RANGE_MESSAGE);

    const salahValues = req.entries.map((entry) => salahToDb(entry.salah));
    const existing = await this.prisma.iqamaEntry.findMany({
      where: { organizationId: req.organizationId, date: req.effectiveDate, salah: { in: salahValues } },
    });
    const byValue = new Map(existing.map((row) => [row.salah, row]));
    const saved = [];
    for (const item of req.entries) {
      const value = salahToDb(item.salah);
      const data = {
        time: timeOnlyToDb(item.time),
        offsetMinutes: item.offsetMinutes,
        note: item.note?.trim() ? item.note.trim() : null,
        updatedAtUtc: new Date(),
      };
      const current = byValue.get(value);
      saved.push(
        current
          ? await this.prisma.iqamaEntry.update({ where: { id: current.id }, data })
          : await this.prisma.iqamaEntry.create({ data: { id: randomUUID(), organizationId: req.organizationId, date: req.effectiveDate, salah: value, ...data } }),
      );
    }
    return saved.map(toIqamaEntryJson).sort((a, b) => salahOrder(a.salah) - salahOrder(b.salah));
  }

  @Post()
  @AuthorizeRoles('Admin', 'Editor')
  async create(@CurrentUser() user: SessionUser | null, @Body(validated(iqamaUpsertSchema)) req: IqamaUpsertRequest, @Res({ passthrough: true }) response: Response): Promise<IqamaEntryJson> {
    if (!canAccessOrganization(user, req.organizationId)) throw forbidden();
    if (req.offsetMinutes !== null && (req.offsetMinutes < 0 || req.offsetMinutes > 180)) throw badRequestText(OFFSET_RANGE_MESSAGE);
    const duplicate = await this.prisma.iqamaEntry.findFirst({ where: { organizationId: req.organizationId, date: req.date, salah: salahToDb(req.salah) } });
    if (duplicate) throw conflict('An Iqama entry already exists for that prayer and effective date.');
    const created = await this.prisma.iqamaEntry.create({
      data: {
        id: randomUUID(),
        organizationId: req.organizationId,
        date: req.date,
        salah: salahToDb(req.salah),
        time: timeOnlyToDb(req.time),
        note: req.note,
        offsetMinutes: req.offsetMinutes,
        updatedAtUtc: new Date(),
      },
    });
    response.status(201).setHeader('Location', `/api/v1/iqama/${created.id}`);
    return toIqamaEntryJson(created);
  }

  @Put(':id')
  @AuthorizeRoles('Admin', 'Editor')
  @HttpCode(204)
  async update(@CurrentUser() user: SessionUser | null, @Param('id') id: string, @Body(validated(iqamaUpsertSchema)) req: IqamaUpsertRequest): Promise<void> {
    const existing = isGuid(id) ? await this.prisma.iqamaEntry.findFirst({ where: { id } }) : null;
    if (!existing) throw notFound();
    if (!canAccessOrganization(user, existing.organizationId) || req.organizationId.toLowerCase() !== existing.organizationId.toLowerCase()) throw forbidden();
    if (req.offsetMinutes !== null && (req.offsetMinutes < 0 || req.offsetMinutes > 180)) throw badRequestText(OFFSET_RANGE_MESSAGE);
    await this.prisma.iqamaEntry.update({
      where: { id: existing.id },
      data: { date: req.date, salah: salahToDb(req.salah), time: timeOnlyToDb(req.time), note: req.note, offsetMinutes: req.offsetMinutes, updatedAtUtc: new Date() },
    });
  }

  @Delete(':id')
  @AuthorizeRoles('Admin', 'Editor')
  @HttpCode(204)
  async remove(@CurrentUser() user: SessionUser | null, @Param('id') id: string): Promise<void> {
    const existing = isGuid(id) ? await this.prisma.iqamaEntry.findFirst({ where: { id } }) : null;
    if (!existing) throw notFound();
    if (!canAccessOrganization(user, existing.organizationId)) throw forbidden();
    await this.prisma.iqamaEntry.delete({ where: { id: existing.id } });
  }
}

function requireGuid(value: string | undefined): string {
  return value && isGuid(value) ? value : '00000000-0000-0000-0000-000000000000';
}
